src/page-storage/store-page.js

```diff
@@ -1,66 +1,97 @@
-    return {finalPage}
+    return finalPage
 }
 
-// Tries to reidentify the page in the tab, or creates a new page doc for it.
+// Tries to reidentify the page in tab, or creates a new page doc for it.
 // Either way, it returns the page doc. If a new page was created, it starts
-// page analysis and then deduplication, and also returns a promise
-// (finalPagePromise) of the page doc resulting from these steps.
-export async function reidentifyOrStorePage({tabId, url}) {
+// page analysis and then deduplication, and also returns an async function
+// (fetchFinalPage) returning the page doc resulting from these steps.
+export async function reidentifyOrStorePage({ tabId, url }) {
     // Find pages we know that had the same URL.
-    const samePageCandidates = (await findPagesByUrl({url})).rows.map(row => row.doc)
+    const samePageCandidates = (await findPagesByUrl({ url })).rows.map(row => row.doc)
 
     // Check if we can already tell in advance that this is the same page.
-    const reusablePage = await tryReidentifyPage({tabId, url, samePageCandidates})
+    const reusablePage = await tryReidentifyPage({ tabId, url, samePageCandidates })
 
     if (reusablePage) {
         // Equality is known (or assumed) in advance. Reuse the old page as is.
-        return {page: reusablePage}
+        return { page: reusablePage }
     } else {
-        // Create a new page doc in the database.
-        const page = await createPageStub({url})
-
-        // Start analysis and (possibly) deduplication, but do not wait for it.
-        const finalPagePromise = analyseAndTryDedupePage({
-            tabId,
-            page,
-            samePageCandidates,
-        })
-
-        // Return the page stub, and a promise of the analysed & deduped page.
-        return {page, finalPagePromise}
+        // Create a new page doc stub in the database.
+        const page = await storePage(createPage({ url }))
+
+        // Set up fetching function to run anaylsis + fetching logic on-demand
+        const fetchFinalPage = () => analyseAndTryDedupePage({ tabId, page, samePageCandidates })
+
+        // Return the page stub, and a function to get the analysed & deduped page.
+        return { page, fetchFinalPage }
+    }
+}
+
+/**
+ * Given a URL, attempts to perform all logic needed to store a full page document.
+ * @param {string} url The URL to attempt to store a page document for.
+ * @returns {any} Page document that was created (or found, if reusable/deduped).
+ */
+export async function storePageFromUrl(url) {
+    // Find pages we know that had the same URL.
+    const samePageCandidates = (await findPagesByUrl({ url })).rows.map(row => row.doc)
+
+    // Check if we can already tell in advance that this is the same page.
+    const reusablePage = await tryReidentifyPage({ url, samePageCandidates })
+
+    // Equality is known (or assumed) in advance. Reuse the old page as is.
+    if (reusablePage) {
+        return reusablePage
     }
```

Comment:
No rush with replies. 

Alright so by breaking it into separate "browser history import" and "fetch&analysing pages" processes, in your mind, what does each process consist of? Is "browser history import" just the `browser.history.search()` call to get `HistoryItem`s, along with the creation of `page` stub docs for each of them? If so, then yeah the visit docs should be able to be done in that stage (cannot do in batch stage as processing order is not guaranteed, by design).

> It makes most sense to me to not rely too much on the fetching process; e.g. to allow deferring the fetching to another time.

This makes me question when these processes/stages will happen in relation to user actions in the UI. At the moment the UI is based on @oliversauter's mockups, linked to in #65 (implementation now looks like [this](http://imgur.com/a/EXS1b)), and pressing "Start Import" currently starts off the batcher (with the other buttons all essentially affording user-control of the batcher's state), which is already ready-to-run in the background with input URLs gotten from a `browser.history.search()` call on background script start. 

As you have noted, history import process shouldn't take long (compared with batch fetching process), but it's input size dependent, so probably should not just happen on background script load, as I'm playing around with now. It could happen on the "Start Import" press, or even when the user clicks on the "Import" option in options nav, or even have a pre-import view before the estimates view that runs that process when the user wants to proceed into import estimate view. As you brought up the possibility of allowing fetching process to be deferred, did you already have something in mind for when the "browser history import" process should happen relative to UI/user actions?